### DISCUSSION TITLE
Make distinction between queries and mutations optional?

### DISCUSSION THREAD

**Discussion by KATT** [trpc/trpc#2663] [RFC / Ideas]
> Originally posted [here](https://github.com/trpc/trpc/issues/2653#issuecomment-1241548010)

I had an idea the other week of potentially allowing some API that allows this for people who want it.

### Server
```ts
// Instruct tRPC that we don't want the `query`/`mutation` distinction
const t = initTRPC.pureRPC().create();

// Procedures are defined with `.resolve` at the end rather than `.query`/`.mutation`
const appRouter = t.router()
  greeting: t.procedure
   .input(z.string().optional())
   .resolve((opts) => `Hello ${opts.input ?? 'World'}!`),
});

export type AppRouter = typeof appRouter;
```

### Raw client

```ts
// client.ts

const client = createTRPCProxyClient<AppRouter>({
  url: "..."
})

// Will be called with `()` in the client, will always do a `POST` request for HTTP
appRouter.greeting();
```

### React

```ts
const trpc = createTRPCReact<AppRouter>();

// both of these are OK
trpc.greeting.useQuery();
trpc.greeting.useMutation();
```

## Notes

- Unsure when I will prioritize exploring this futher, if ever
- [A vast majority](https://twitter.com/alexdotjs/status/1528854148011130880) of people might seem to like the query/mutation distinction

**Comment by cseitz** [comment]
This is something I've been interested in ever since I started using TRPC.

I agree that the intentional distinction between queries and mutations is an important aspect; but in some cases when I'm writing queries or mutations it may be overly verbose.

For example, `trpc.users.findOne.query({ id: 1234 })` would, in my opinion, be cleaner to write as `trpc.users.get({ id: 1234 })`. 
The distinction of it being a query is in the name of the procedure, which in some cases is enough to identify what the call is going to do. 
Same goes for `trpc.tickets.update({ id: 1234, status: TicketStatus.Open })` has to be `trpc.tickets.update.mutation({ ... })` right now.

Opting out of the query/mutation distinction is certainly inviting ambiguity; but usually I don't have enough procedures where this would ever be a problem. (especially because I use JSDoc comments on all my procedures, so they're well-documented when I hover over them in the editor).

I think that being able to manually opt-out of query/mutation distinction would be a small quality-of-life change for casual use-cases of TRPC. I mean I literally use TRPC for every project I work on now; and most are just personal projects! 

Being able to just go with `resolve` makes TRPC feel even more like what its name implies; TypeSafe Remote Procedure Calls! Without specifying query or mutation, I get to write my API calls as if I was actually calling the function! Sure its just a single word difference, but for some reason it just feels more... natural?


But yeah, knowing that you guys already have a lot on your plate; it really isn't worth the development time.

Nevertheless, if you ever got a chance, I'd love to see a feature such as this!

**Comment by capaj** [comment]
I would keep the distinction. React-query keeps it too and trpc client is built on top of it. I think it is crucial to be able to distinguish just glancing over the code whether a remote call mutates server state or not.
